src/Common/src/System/Text/DBCSDecoder.cs

```diff
@@ -103,7 +104,7 @@ public unsafe override int GetCharCount(byte* bytes, int count, bool flush)
             if (count < 0)
                 throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedNonNegNum);
 
-            bool excludeLastByte = count > 0 || (!flush && IsLastByteALeadByte(bytes, count));
+            bool excludeLastByte = count > 0 && !flush && IsLastByteALeadByte(bytes, count);
```

Comment:
by the way, before fixing the condition the code should be working too but it was always handling last byte as lead byte which will not change the behavior but would slow a little the perf.